ATB/DataAccess/Repositories/Implementations/CsvBookingRepository.cs

```diff
@@ -0,0 +1,85 @@
+using DataAccess.CsvHelperService;
+using DataAccess.Enums;
+using DataAccess.Models;
+using DataAccess.Repositories.Interfaces;
+using DataAccess.SearchCriteria;
+
+namespace DataAccess.Repositories.Implementations;
+
+public class CsvBookingRepository : IBookingRepository
+{
+    private readonly string _pathToCsv;
+    private readonly ICsvFileService<Booking> _csvFileService;
+    private List<Booking> _bookingsCache;
+    private bool _isCacheInitialized = false;
+
+    public CsvBookingRepository(ICsvFileService<Booking> csvFileService, string pathToCsv)
+    {
+        _csvFileService = csvFileService;
+        _pathToCsv = pathToCsv;
+    }
+
+    private async Task InitializeCacheAsync()
+    {
+        if (!_isCacheInitialized)
+        {
+            _bookingsCache = await _csvFileService.ReadFromCsvAsync(_pathToCsv);
+            _isCacheInitialized = true;
+        }
+    }
+
+    public async Task<IEnumerable<Booking>> GetAllAsync()
+    {
+        await InitializeCacheAsync();
+        return _bookingsCache;
+    }
+
+    public async Task<Booking?> GetByIdAsync(Guid id)
+    {
+        await InitializeCacheAsync();
+        return _bookingsCache.FirstOrDefault(b => b.Id == id);
+    }
+
+    public async Task AddAsync(Booking booking)
+    {
+        await InitializeCacheAsync();
+        _bookingsCache.Add(booking);
+        await _csvFileService.WriteToCsvAsync(_pathToCsv, _bookingsCache);
+    }
+
+    public async Task UpdateAsync(Booking booking)
+    {
+        await InitializeCacheAsync();
+        var index = _bookingsCache.FindIndex(b => b.Id == booking.Id);
+        if (index != -1)
+        {
+            _bookingsCache[index] = booking;
+            await _csvFileService.WriteToCsvAsync(_pathToCsv, _bookingsCache);
```

Comment:
Have you thought of the consequences of these 2 lines being not atomic? what if line 56 succeeded and the cache is updated, but line 57 failed due to some error, our system now is in a data corruption state, the cache is updated but the value in our db/csv is not updated, they are out of sync.
What about implementing a simple mechanism to invalidate/remove the cache entry in case of updating and the next time we retrieve it from db/csv we cache it again so we get the last updated version of the entity?